src/bartiq/analysis/rewriters/utils.py

```diff
@@ -61,12 +85,47 @@ def from_string(cls, assumption_string: str) -> Self:
         return cls(*_unpack_assumption(assumption_string))
 
 
+@dataclass(frozen=True)
+class Substitution(Instruction):
+    """Substitute a symbol with an expression."""
+
+    symbol_or_expr: str
+    replacement: str
+    backend: SymbolicBackend
+    wild: tuple[str, ...] = ()
+
+    def __post_init__(self):
+        object.__setattr__(self, "wild", _get_wild_characters(self.symbol_or_expr))
+
+    def _get_linked_parameters(self) -> dict[str, Iterable[str]]:
+        if self.wild:
+            raise NotImplementedError("Unable to derive connections between wild characters and expression symbols.")
+
+        def free_symbols_in_expr(expr: str) -> Iterable[str]:
+            return self.backend.free_symbols(self.backend.as_expression(expr))
+
+        return {
+            x: y
+            for x in free_symbols_in_expr(self.replacement)
+            if x not in (y := free_symbols_in_expr(self.symbol_or_expr))
+        }
+
+
+def _get_wild_characters(expression: str) -> tuple[str, ...]:
+    return tuple(re.findall(r"\$([a-zA-Z_][a-zA-Z0-9_]*)", expression))
+
+
 def _get_properties(comparator: str, reference_value: int | float) -> dict[str, bool | None]:
     """Derive properties of an assumption.
 
-    At present this only detects positivity/negativity.
-
-    If the properties are unknowable due to lack of information, they are None.
+    At present this only detects positivity/negativity. The two are calculated independently,
```

Comment:
[nitpick] `this only detects positivity/negativity` -> `this only detects positivity/negativity of a given symbol/expression` ?